### DISCUSSION TITLE
2.0.0 Feedback & Ideas

### DISCUSSION THREAD

**Discussion by Cramertech** [noah-nuebling/mac-mouse-fix#82] [Other]
Hey Noah! Just thought maybe some input could be appreciated on the beta of the new 2.0.0. Maybe others will join in and offer their thoughts too.

I really like the automatic ability to hold the cmd key and use the mouse scroll wheel to zoom in with Safari. I had been looking and looking for this. I felt like it had to be built in and I just couldn't find it. Finally someone pointed out in a Stack Exchange topic that it's not a part of MacOS or Safari by default. Obviously I have been spoiled by using a trackpad or magic mouse at some point and must have assumed it was native with normal mice too.

I think you're on to something with the whole ability to add extra mouse buttons. Steer Mouse offers this and costs like $20, so it's really nice that your app is offering it free. I only have buttons Middle, and 4 and 5, so I can't really test that beyond what was already provided before (tried on the top button, but discovered it is just a sensitivity toggle). I like that I can assign any keyboard shortcut to any of the buttons, but I found it to behave oddly when I tried to enter cmd and plus (⌘+). Obviously your built-in cmd+scroll feature negates the need for it, but at any rate, I tried and it just did weird stuff when I tried using those buttons (maybe a bug?).

One thing I did notice was that the Click and Drag Middle Button option only shows Mission Control & Spaces. Would be nice to allow that to do any of the custom functions including keyboard shortcuts.

Also maybe is it possible to have all the available buttons offer the option of sending just a standard "Button X" system event sort of thing? This could be useful for mice that don't have driver software to support the additional buttons when trying to do custom things inside of BetterTouchTool.

Thanks!

**Comment by noah-nuebling** [comment]
Hey Jason, thanks so much for your input!

I'm glad you like the scroll-zooming, I think it's really cool too, and I also always felt like it should totally be a native feature.

That's interesting that the zooming shortcut doesn't work for you. Using a standard English keyboard layout the shortcut works for me but it shows up as (⌘=) instead of (⌘+) in the app for me. 

![Screenshot 2021-04-11 at 23 26 21](https://user-images.githubusercontent.com/40808343/114321850-55c0fd00-9b1d-11eb-8632-c1cee4c3d1ac.png)

Are you maybe using a different keyboard layout or do you have any idea what else might be causing it? 

Assigning other actions to click and drag might be really useful. That's a really interesting idea but it would be a lot of work to implement. That's definitely on the radar though. 
However,  you can also assign more actions to a button, by using double click, triple click, click and hold, keyboard modifiers, etc. which should cover some of the use cases.

Sending standard button events shouldn't be super hard to implement, but I didn't prioritise it up until now, because there weren't many use cases that I could think of, and making a button "properly" act like another button (so it also works for click and hold, and click and drag) isn't something Mac Mouse Fix can do currently for a bunch of technical reasons. 
If you have time it would be super cool if you could you describe that use case with BetterTouchTool you were thinking of in more detail. Thanks!

**Reply by Cramertech** [reply]
Sure.

Basically BetterTouchTool can do actions based on a mouse's button inputs, but natively most mice don't have support for those buttons on MacOS. If your tool clearly has access to those buttons, then all we would need instead of a driver (and I usually hate 1st party driver software which is usually bloated), is for your app to just pass the button click system events (button down, button up) for whichever button was pressed.

For example if I press button 4, then Button 4 down & then Button 4 up would both be sent to the system. Therefore BetterTouchTool would be able to detect these and then I could trigger custom functionality with them.

Essentially this would just be a pass-through feature so that various apps which have support for extra mouse buttons will be able to actually make use of them. To test this while you develop the feature, you could install a trial of BetterTouchTool if you like. I'd be happy to test it as well.

----

As for the shortcut on MB's 4 and 5 with zooming hotkeys, I guess it was a matter of I was trying to use shift ⇪ with it, but it doesn't like that actually. I saw your screenshot and did it without shift and it works fine. It's weird because the actual hotkeys are supposed to be ⌘+ and ⌘- ...but the one for zoom-in defaults to an = sign without holding shift. It's just weird that by doing what you kind of would expect, it doesn't behave right - and that's a MacOS/Safari issue. No biggie at all though since now I get it. Just might end up needing to be explained at some point potentially.

**Reply by noah-nuebling** [reply]
Hey Cramertech, sorry for the late reply, and thanks for your answer!

> Basically BetterTouchTool can do actions based on a mouse's button inputs...

That's an interesting use-case that I hadn't considered, and a very good reason to have a pass-through feature. You've put it on the map for me with that elaboration. The problem is that Mac Mouse Fix actually doesn't have any special code to support devices which send non-standard events, so I don't think there would be cases where Mac Mouse Fix can detect a button press of some mouse, and BTT can't. I'd love to have such device specific support at some point though. 
But even if we have that, implementing a pass-through feature within the framework of the new remapping engine in MMF 2, would be much harder than you might think. More on that below if you're interested.

So while this is on my radar now, please don't expect anything soon, or maybe ever, as it's not a super high priority at the moment and would probably be hard to implement well.

Thanks for offering your help to test the pass-through feature! I appreciate that.

> As for the shortcut on MB's 4 and 5 with zooming hotkeys...

Yeah that is quite unintuitive, I understand your confusion. Good that everything's cleared up now. Thanks for getting back about it!

---

__If you are interested here's an overview as to why it would be hard to implement a pass-through feature within the framework of the new remapping engine of MMF 2__

In the new remapping engine, mouse buttons can have an effect in 2 different ways. As triggers - producing actions directly - or as modifiers - changing what effects other triggers produce.

For various reasons, whatever actions a button produces as a trigger directly have to be 'oneShot' that is they have to be one and done. The action has to fire at a certain point in time and then be over immediately. This restriction is necessary for things to work properly if you have several actions assigned to a mouse button. This makes it possible to have different actions when clicking, clicking and holding, double clicking, etc.. without those actions interfering.

Now, simulating a standard button click, when a button which sends some non-standard event is clicked doesn't fit into this framework very well, because it's not one and done, it's extended over time. There should be a mouse-down event, when you click the mouse button, and a mouse-up event when you release it.

There are several possible ways I can think of to get around this, but all of them would bring some caveats and/or be a lot of work to implement.

**Comment by hezral** [comment]
Logitech extra buttons

I have a logitech m337, with a navigation button just above the scroll wheel. I tried to add it but seems like its not detecting it. 
How do i check if Mac Mouse Fix is able to detect the buttons?

**Reply by noah-nuebling** [reply]
Hi @hezral, there are many mice that have buttons which Mac Mouse Fix can’t detect because they send non-standard events. This is very common with Logitech mice, so I suspect that’s what’s going on for you.

Normally you can add a button like so:

1. Click the “+” button to the left of the  “-“ button
2. Place the mouse pointer over the field with the “+” in the middle
3. Click / Hold / Double Click / ... the mouse button you’d like to add

If that doesn’t work then Mac Mouse Fix probably doesn’t detect that mouse button.

Alternatively you could check out SteerMouse or use Logitech Options. Those should work with your extra button. USB Overdrive might work as well.

**Comment by hezral** [comment]
Thanks. Yeah i suspected that but didn't want to put a separate app like
Logitech Options. I'll try what fits.

**Comment by huszarrichard** [comment]
Hi, unfortunately doesn't work smooth scrolling and invert direction. This happened today, yesterday was ok. On Version 2.0.0 Beta 8 and Beta 9 too. 
![image](https://user-images.githubusercontent.com/17738461/117438296-aca9ce80-af31-11eb-823d-d7d56b93803d.png)

**Reply by noah-nuebling** [reply]
Hmm that's weird I don't know what might be causing it. 

Here are some things you could test to help us get to the bottom of this:

- Test if [1.0.4](https://github.com/noah-nuebling/mac-mouse-fix/releases/tag/1.0.4) works, or if it has the same problems
- File a [Bug Report](https://noah-nuebling.github.io/mac-mouse-fix-feedback-assistant/?type=bug-report) and attach some console logs that were recorded while scrolling. The process for this this is described on the Bug Report page but it can still be a little confusing. If you have questions just let me know!
- Think about if there is anything that changed since the last time scrolling worked in Mac Mouse Fix. Have you maybe installed some new software, or are you now logged into you computer with several user accounts at once?

Thanks for your contribution @huszarrichard!

**Comment by Spillmaker** [comment]
Hi, would it be possible to add support for the "Gesture"-button on the MX-master series from Logi(tech)?

When i try to add the mouse-button with the plus-icon, it doesn't seem to register the click.

This tool would be perfect for my usecase if it would be possible to use the Gesture-button the same way as the middle button click and drag-feature has. (Pressing the gesture-button and dragging to swap spaces), and Logitech seems to not be too eager to update their Options-app for Apple Sillicon.

**Reply by noah-nuebling** [reply]
Hi @Spillmaker, thanks for your request!

I really want to add ful support for Logitech Mice, but it's a lot of work, so it's unfortunately not coming very soon. I'm on it though!

**Comment by wgdavies** [comment]
Great app, @noah-nuebling ! Thanks for the effort and making it available! Milkshake coming soon. :) 

I am trying this out as a lightweight version of Smooze, which I don't mind paying for but it's a lot to purchase yet another license (for my fourth device now :P) just to add _one_ piece of functionality that I occasionally need:
I have a prized HP M-UD43 (I didn't make up the designation, honest!) 3-button mouse. The third button works great for the standard Unix/X Terminal pasting, closing browser tabs, &c but the one thing it _doesn't_ do is click'n'drag window contents. 

That's it. That's all I need to do. So even this beautiful app is total overkill! 

Any suggestions?

**Reply by noah-nuebling** [reply]
Hey @wgdavies, thanks for your feedback, and sorry for the late reply! I'm not sure Mac Mouse Fix can do that, or maybe I'm misunderstanding? Which option in Smooze/MMF are you using to achieve this? 

If it's not in MMF already, maybe I can implement something like this in the future.

Note that Mac Mouse Fix will also be paid in the future, but it will be pretty cheap and you'll be able to use one license on all your devices.
You can read more about that [here](https://github.com/noah-nuebling/mac-mouse-fix/releases/tag/3.0.0-Beta-1.1).

**Comment by neblina-of-the-night** [comment]
Hi! Nice App!!! congrats from Brazil!

Idea: New Option for open application with mouse click. Ex.: Open Finder or Safari with Middle Button or Button 4.

**Reply by noah-nuebling** [reply]
Hi @neblina-of-the-night, thanks for your idea and for the nice words!

I must admit, I don't totally see how it's useful to open apps using a mouse button. How/why would you use the feature?

One thing you could already do is use some other software like [Automator](https://www.computerhope.com/issues/ch002051.htm) to map a keyboard shortcut to opening an app and then use MMF to trigger that keyboard shortcut.

Hope that helps a little. Let me know if you have more thoughts or questions to share!
